sdk/core/azure-core/azure/core/async_paging.py

```diff
@@ -85,26 +88,45 @@ def __init__(
          list of ReturnType
         :param str continuation_token: The continuation token needed by get_next
         """
-        self._get_next = get_next
-        self._extract_data = extract_data
+        self._initial_request = kwargs.pop("initial_request", None)
+        self._initial_response = kwargs.pop("initial_response", None)
+        if self._initial_response:
+            self._initial_request = self._initial_response.http_response.request
+
+        if get_next or extract_data:
+            if paging_method:
+                raise ValueError(
+                    "You can't pass in both a paging method and a callback for get_next or extract_data. "
+                    "We recomment you only pass in a paging method, since passing in callbacks is legacy."
+                )
+            self._paging_method = _LegacyPagingMethod(get_next, extract_data)
+        else:
+            if not self._initial_request and not self._initial_response:
+                raise ValueError(
+                    "You must either supply the initial request the paging method must call, or provide "
+                    "the initial response"
+                )
+            self._paging_method = paging_method
+            self._paging_method.initialize(**kwargs)
         self.continuation_token = continuation_token
-        self._did_a_call_already = False
-        self._response = None
-        self._current_page = None
+        self._response = None  # type: Optional[ResponseType]
+        self._current_page = None  # type: Optional[Iterable[ReturnType]]
 
     async def __anext__(self):
-        if self.continuation_token is None and self._did_a_call_already:
+        if self._paging_method.finished(self.continuation_token):
             raise StopAsyncIteration("End of paging")
         try:
-            self._response = await self._get_next(self.continuation_token)
+            try:
+                self._response = await self._paging_method.get_page(self.continuation_token, self._initial_request)
+            except TypeError:
+                # legacy doesn't support passing initial request into get_page
+                self._response = await self._paging_method.get_page(self.continuation_token)
```

Comment:
Is there a reason that the TypeError leaks out of the legacy methods get_page? Who else calls the method? If it is only intended to be called from here, and we always ignore the failure, then the method itself might as well ignore it...

Reply:
...which incidentally would get rid of the nested `try's :)

Reply:
In the new design, we define a `get_page` on the paging method, that takes in a cont token like it used to, but also takes in the initial request so it's more readily available for users to play with when they're overriding the paging method class. So, if we pass in initial request to the legacy callbacks people used to send, we get a `TypeError` if we pass in the initial request, which is why I catch it. However, after writing this whole chunk of response, I'm thinking I can deal with this in the `_LegacyPagingMethod` and get rid of the nested trys